pkg/vault/contracts/ProtocolFeePercentagesProvider.sol

```diff
@@ -0,0 +1,143 @@
+        (_maxProtocolSwapFeePercentage, _maxProtocolYieldFeePercentage) = protocolFeeController
+            .getMaximumProtocolFeePercentages();
+    }
+
+    /// @inheritdoc IProtocolFeePercentagesProvider
+    function getProtocolFeeController() external view returns (IProtocolFeeController) {
+        return _protocolFeeController;
+    }
+
+    /// @inheritdoc IProtocolFeePercentagesProvider
+    function getFactorySpecificProtocolFeePercentages(
+        address factory
+    ) external view returns (uint256 protocolSwapFeePercentage, uint256 protocolYieldFeePercentage) {
+        FactoryProtocolFees memory factoryFees = _getValidatedProtocolFees(factory);
+
+        protocolSwapFeePercentage = factoryFees.protocolSwapFeePercentage;
+        protocolYieldFeePercentage = factoryFees.protocolYieldFeePercentage;
+    }
+
+    /// @inheritdoc IProtocolFeePercentagesProvider
+    function setFactorySpecificProtocolFeePercentages(
+        address factory,
+        uint256 protocolSwapFeePercentage,
+        uint256 protocolYieldFeePercentage
+    ) external authenticate {
+        // Validate the fee percentages; don't store values that the `ProtocolFeeCollector` will reject.
+        if (protocolSwapFeePercentage > _maxProtocolSwapFeePercentage) {
+            revert IProtocolFeeController.ProtocolSwapFeePercentageTooHigh();
+        }
+
+        if (protocolYieldFeePercentage > _maxProtocolYieldFeePercentage) {
+            revert IProtocolFeeController.ProtocolYieldFeePercentageTooHigh();
+        }
+
+        // Ensure precision checks will pass.
+        _protocolFeeController.ensureValidPrecision(protocolSwapFeePercentage);
+        _protocolFeeController.ensureValidPrecision(protocolYieldFeePercentage);
+
+        // Best effort check that `factory` is the address of an IBasePoolFactory.
+        bool poolFromFactory = IBasePoolFactory(factory).isPoolFromFactory(address(0));
+        if (poolFromFactory) {
+            revert InvalidFactory(factory);
+        }
+
+        // Store the default fee percentages, and mark the factory as registered.
+        _factoryDefaultFeePercentages[IBasePoolFactory(factory)] = FactoryProtocolFees({
+            protocolSwapFeePercentage: protocolSwapFeePercentage.toUint64(),
+            protocolYieldFeePercentage: protocolYieldFeePercentage.toUint64(),
+            isFactoryRegistered: true
+        });
+
+        emit FactorySpecificProtocolFeePercentagesSet(factory, protocolSwapFeePercentage, protocolYieldFeePercentage);
+    }
+
+    /// @inheritdoc IProtocolFeePercentagesProvider
+    function setProtocolFeePercentagesForPools(address factory, address[] memory pools) external {
+        FactoryProtocolFees memory factoryFees = _getValidatedProtocolFees(factory);
+
+        for (uint256 i = 0; i < pools.length; ++i) {
+            address currentPool = pools[i];
+
+            if (IBasePoolFactory(factory).isPoolFromFactory(currentPool) == false) {
```

Comment:
Technically if you set the fee percentage in a contract that just returns `true` for any address that is not 0, you can use that 'factory' to set any pool fees.

I think there's not much way around it. I don't see much value in splitting `setFactorySpecificProtocolFeePercentages` into 2 steps (one to register the factory, and another one to set the value) as in practice it'd be pretty much the same.